### PR TITLE
[NativeAnimated][Android] Add support for animated events

### PR DESCRIPTION
**Pull request by janicduplessis**
This adds support for `Animated.event` driven natively. This is WIP and would like feedback on how this is implemented.

At the moment, it works by providing a mapping between a view tag, an event name, an event path and an animated value when a view has a prop with a `AnimatedEvent` object. Then we can hook into `EventDispatcher`, check for events that target our view + event name and update the animated value using the event path.

For now it works with the onScroll event but it should be generic enough to work with anything.

### CODE REVIEW

**Comment by facebook-github-bot-9** [comment]
By analyzing the blame information on this pull request, we identified @kmagiera and @ericvicenti to be potential reviewers.

**Comment by janicduplessis** [comment]
cc @brentvatne

**Comment by janicduplessis** [comment]
Other thought:
- Should Animated Events be represented somehow as a node in the graph? I went with a separate data structure to just get it working.
- We need to map event names properly. On Android events have different names than in JS, ie topScroll for the onScroll prop. There is a mapping of that in UIManager so it should be possible to use that to convert it properly.

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 1]
```diff
@@ -1462,6 +1466,30 @@ function createAnimatedComponent(Component: any): any {
...
+      // `ScrollResponder.Mixin`.
+      const ref = this.refs[refName].getScrollableNode ?
+        this.refs[refName].getScrollableNode() :
+        this.refs[refName];
+
+      for (let key in nextProps) {
+        const prop = nextProps[key];
+        if (prop instanceof AnimatedEvent && prop.__isNative) {
+          // TODO: Map event names using the map in UIManager.
+          if (key === 'onScroll') {
```
Comment (brentvatne) [inline Libraries/Animated/src/AnimatedImplementation.js]: Which props aside from onScroll do you think we might want to be able to bind natively to with Animated.event?

Comment (brentvatne) [inline Libraries/Animated/src/AnimatedImplementation.js]: Actually pan responders would be a good example of these kinds of events

**Comment by brentvatne** [comment]
This is so exciting!

> Should Animated Events be represented somehow as a node in the graph? I went with a separate data structure to just get it working.

Not sure about this one, will leave it to someone else to comment on.

> Is there another way to attach the event to the view? Right now if the view uses a native driven event it need to be an animated view, ie: Animated.ScrollView.

Another option would be to just make ScrollView take care of attaching the event, but this doesn't work for other views and leaks the responsibility for this outside of the Animated codebase.

We need to know:
- what event to watch for (name of the prop -> event mapping)
- react tag for the view
- when to detach the event

I'm sure that there are ways we could work around this but I can't think of any solid ones,  so I'm for `Animated.ScrollView` here.

cc @vjeux @sahrens

**Comment by foghina** [comment]
cc @kmagiera

File: ReactAndroid/src/main/java/com/facebook/react/animated/PropsAnimatedNode.java [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 package com.facebook.react.animated;
 
+import android.util.Log;
```
Comment (kmagiera) [inline ReactAndroid/src/main/java/com/facebook/react/animated/PropsAnimatedNode.java]: revert this

File: ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java [hunk 3]
```diff
@@ -38,16 +45,18 @@
...
   private final ArrayList<AnimationDriver> mActiveAnimations = new ArrayList<>();
   private final ArrayList<AnimatedNode> mUpdatedNodes = new ArrayList<>();
+  private final Map<String, EventAnimationDriver> mEventDrivers = new HashMap<>();
   private final UIImplementation mUIImplementation;
   private int mAnimatedGraphBFSColor = 0;
 
-  public NativeAnimatedNodesManager(UIImplementation uiImplementation) {
+  public NativeAnimatedNodesManager(UIImplementation uiImplementation, EventDispatcher eventDispatcher) {
     mUIImplementation = uiImplementation;
+    eventDispatcher.addListener(this);
```
Comment (kmagiera) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: I think we should not dispatch event to JS in the case when it is handled by the native animated driver, this would generate an unnecessary traffic over the bridge

File: ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java [hunk 4]
```diff
@@ -230,6 +239,42 @@ public void disconnectAnimatedNodeFromView(int animatedNodeTag, int viewTag) {
...
+    mEventDrivers.put(viewTag + eventName, event);
+  }
+
+  public void removeAnimatedEventFromView(int viewTag, String eventName) {
+    mEventDrivers.remove(viewTag + eventName);
+  }
+
+  @Override
+  public void onEventDispatched(Event event) {
+    if (!mEventDrivers.isEmpty() && mEventDrivers.containsKey(event.getViewTag() + event.getEventName())) {
```
Comment (kmagiera) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: checking emptiness is unnecessary. You can also do a `get` before that if which will save one hashmap lookup as well as one string allocation for concatenating tag with the name

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 1 above)
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: @brentvatne this approach will only work for "direct" events. That is because touch handling has a bubbling logic written in JS and we cannot determine which view is going to be responsible for handling the touch without running some JS code

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 5]
```diff
@@ -1850,13 +1962,19 @@ var event = function(
...
+  }
+}
+
+var event = function(
+  argMapping: Array<?Mapping>,
+  config?: ?EventConfig,
+): any {
+  const animatedEvent = new AnimatedEvent(argMapping, config);
+  if (animatedEvent.__isNative) {
+    return animatedEvent;
```
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: we should perhaps validate that argMapping has just `nativeEvent` key present (that's the data provided from the native part). There are many cases when you may want to use synthetic params (e.g. with pan responder) and in case someone try to set synthetic values mapping for natively driven value we should display a warning

**Comment by facebook-github-bot-1** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 6]
```diff
@@ -1505,21 +1509,58 @@ function createAnimatedComponent(Component: any): any {
...
     }
 
     componentDidMount() {
       this._propsAnimated.setNativeView(this._component);
+
+      this._attachNativeEvents(this.props);
+    }
+
+    _attachNativeEvents(newProps) {
+      if (newProps !== this.props) {
```
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: This can perhaps be optimised so that we don't reattach native props every time when non relevant property gets changed. Also, I'm not sure if this statement can even evaluate to true (if props hasn't been updated `componentWillReceiveProps` will not be triggered)

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 7]
```diff
@@ -1880,21 +1921,106 @@ var stagger = function(
...
+    // Find animated values in `argMapping` and create an array representing their
+    // key path inside the `nativeEvent` object. Ex.: ['contentOffset', 'x'].
+    const eventMappings = [];
+
+    const traverse = (value, path: Array<any>) => {
+      if (value instanceof AnimatedValue) {
+        value.__makeNative();
+
+        invariant(
+          path.indexOf('nativeEvent') >= 0,
```
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: I think we should expect nativeEvent to always be the first element on the path, no?

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 8]
```diff
@@ -1880,21 +1921,106 @@ var stagger = function(
...
+          'Native driven events only support animated values contained inside `nativeEvent`.'
+        );
+        const nativeEventPath = path.slice(path.indexOf('nativeEvent') + 1);
+
+        eventMappings.push({
+          nativeEventPath,
+          animatedValueTag: value.__getNativeTag(),
+        });
+        return;
```
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: can you please refactor this to `if { } else { }` it's difficult to notice the return statement here and this type of construct may lead to bugs in the future

File: ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java (hunk 3 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: I think we should still send events to JS since some component could do some JS logic in their callback and there is also the event listener that we would still need to call into js for https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L1887

File: ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java (hunk 4 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: I changed this code a bit but I added the empty check to fast path the case where there is no events to avoid the string concatenation.

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 6 above)
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: I think the common case is to create a new event in every render so it should be fine to reattach everytime, I'll take another look see if I can improve it though.
``` js
<ScrollView
  onScroll={Animated.event({...})
/>
```

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request.

**Comment by janicduplessis** [comment]
Cleanup + added tests, this should be ready for final review.

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 9]
```diff
@@ -1908,17 +2030,23 @@ var event = function(
...
-      config.listener.apply(null, args);
-    }
-  };
+  }
+}
+
+var event = function(
+  argMapping: Array<?Mapping>,
+  config?: ?EventConfig,
+): any {
```
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: Would be good if we could return function from this method. Otherwise it may result in react warnings as `onScroll` is defined to only accept functions (not sure if you're getting them or not).

I see two options to solve this:
1) when event is native, we return a method that will throw an exception when called (handler should never be called as we expect event to be handled in native). Then we can bind some extra arguments to the function object so that we can recognise if the handler is native or not in `AnimatedProps.__getValue` (its a little bit "hacky")

2) Somehow make `Animated.Scroll` wrapper override proptypes so that it allows both function and AnimatedEvent object

File: ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java [hunk 10]
```diff
@@ -234,6 +248,55 @@ public void disconnectAnimatedNodeFromView(int animatedNodeTag, int viewTag) {
...
+
+  @Override
+  public void onEventDispatched(Event event) {
+    // Only support events dispatched from the UI thread.
+    if (!UiThreadUtil.isOnUiThread()) {
+      return;
+    }
+
+    if (!mEventDrivers.isEmpty()) {
+      // If the event has a different name in native convert it to it's JS name.
```
Comment (kmagiera) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: Wouldn't it be better to do the event name conversion at the time when we put the element in `mEventDrivers` and then it won't be necessary to do it here. It should be around same amount of code but this section runs every time we receive the event while the oder one runs once per view

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java [hunk 11]
```diff
@@ -129,6 +130,24 @@ public void dispatchEvent(Event event) {
       // touch event dispatch will hit this codepath, and we simply queue them so that they
       // are dispatched once ReactContext creation completes and JS app is running.
     }
+
+    for (EventDispatcherListener listener : mListeners) {
+      listener.onEventDispatched(event);
```
Comment (kmagiera) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java]: I mentioned that one earlier and not sure if you responded for my comment, but cannot find your response.

I think it would be better if we don't pass events that are handled natively to JS. That event is not going to be utilised on JS side and will just bloat the bridge.

**Comment by kmagiera** [comment]
This looks good, just have few final comments

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 12]
```diff
@@ -1508,21 +1510,58 @@ function createAnimatedComponent(Component: any): any {
...
+        if (prop instanceof AnimatedEvent && prop.__isNative) {
+          prop.__attach(ref, key);
+        }
+      }
+    }
+
+    _detachNativeEvents(props) {
+      // Make sure to get the scrollable node for components that implement
+      // `ScrollResponder.Mixin`.
+      const ref = this._component.getScrollableNode ?
```
Comment (sahrens) [inline Libraries/Animated/src/AnimatedImplementation.js]: Should we make this more generic and not scrolling specific? Have you thought about how to support touchMove/PanResponder events?

**Comment by sahrens** [comment]
Also, this is sick - glad you're making it happen! What's the plan/ETA for iOS?

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 12 above)
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: This only works for direct events, pan responder does add some JS code so we'd have to create some sort of native pan responder if we want to support this. It should be possible to use Animated.event on the native touch move event directly though.

This is just so it works properly with ScrollView which sometimes is wrapped with RefreshControl. We have to use this method to make sure to always get a ref to the native ScrollView.

I havent tested with other events than scrollview onScroll but it is generic enough that it should work, I'll test with touch move on a view today to make sure it works.

**Comment by janicduplessis** [comment]
Haven't looked at iOS yet but it should be possible to implement it the same way. Just wanted to get this reviewed first and settle the JS implementation before starting to work on iOS.

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 9 above)
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: Right now it works with proptypes we return the handler in AnimatedProps (https://github.com/facebook/react-native/pull/9253/files#diff-f048d92ca0be679bc38d38147b311100R1402) so I guess we're doing something like 2).

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java (hunk 11 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java]: Here was my response on the old diff:

I think we should still send events to JS since some component could do some JS logic in their callback and there is also the event listener that we would still need to call into js for https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L1887

File: ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java (hunk 10 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: If I remember correctly the issue with this is that this mapping is a one to many. Many native event name can be mapped to one js event name. I'll double check to make sure it is the case.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java]: I could actually insert into mEventDrivers for each possible mapping.

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 12 above)
Comment (sahrens) [inline Libraries/Animated/src/AnimatedImplementation.js]: Yeah, I was thinking maybe we should change our scroll views to implement getNativeNode or something, and then any component that wants to drive an Animated.Event could just add getNativeNode support and have things work easily?

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 13]
```diff
@@ -1883,21 +1922,104 @@ var stagger = function(
...
+        }
+      }
+    };
+
+    invariant(
+      this._argMapping[0] && this._argMapping[0].nativeEvent,
+      'Native driven events only support animated values contained inside `nativeEvent`.'
+    );
+
+    // Assume that the event containing `nativeEvent` is always the first argument.
```
Comment (sahrens) [inline Libraries/Animated/src/AnimatedImplementation.js]: This shouldn't be hard to change in the future, right?

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java (hunk 11 above)
Comment (kmagiera) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java]: Hey, sorry this link doesn't work for me...

Anyways, I think the whole point of offloading events is to prevent JS thread being bloated by events that we can handle synchronously in native. If there is code that is being triggered "implicitly" we should prevent that from happening. When your intent is to offload events to native than you don't expect to be able to receive them in JS. If you still want to handle events on the JS, than you can use listener param of `Animated.event` like this:
```
onScroll={Animated.event(
   [{nativeEvent: {contentOffset: {x: this._scrollX}}}]
   {listener},          // Optional async listener
 )}
```
In which case we can forward events to JS, but I believe the default should be that we don't do that.
Another way to capture events in JS that will work "out of the box" is using animated value listener like this:
```
onScroll={Animated.event([{nativeEvent: {contentOffset: {x: this._someValue}}}])}
```
And then:
```
this._someValue.addListener(contentOffset => doSomethingWithTheOffset(contentOffset));
```

**Comment by kmagiera** [comment]
@janicduplessis Please see my response here https://github.com/facebook/react-native/pull/9253#discussion_r75342236

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 13 above)
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: We can just change this code to loop through args and look for nativeEvent if this cause an issue in the future because of some event that are different.

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 12 above)
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: We could rename getScrollableNode to getNativeNode so it is more generic and can be use by any component that could have the same issue as scrollview.

**Comment by janicduplessis** [comment]
@kmagiera Updated to not send event to JS if they are handled by native. Let's just not support the animated event listener for now to keep this simpler.

**Comment by satya164** [comment]
> Updated to not send event to JS if they are handled by native

Does this mean that I won't be able to get the updated values with `animatedValue.addListener` when it's updated via `Animated.event` or `animatedValue.setValue`?

Adding a listener parameter to `Animated.Event` doesn't sound very convenient. If I'm passing the animated value to lots of different places and it can be updated from various places, it would be cumbersome to pass a listener everywhere, and remembering to add the listener on consumer side.

My main use case for `addListener` is to get the latest value, since `Animated.Value` doesn't have a method to get the latest value.

**Comment by janicduplessis** [comment]
@satya164 `addListener` will still work as this is handled natively whenever the native animated value changes. The listener parameter for `Animated.event` already existed but I never had to use it since you can get the value you are interested in using a listener on the `Animated.value` anyway.

**Comment by satya164** [comment]
@janicduplessis Oh, so this only affects the listener parameter of `Animated.event`? Doesn't affect me then.

**Comment by janicduplessis** [comment]
@satya164 Yep, exactly

**Comment by facebook-github-bot-6** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/879c13bfd1df34f61092abfcdcaae931ad30b650..861c712028d4b8b5cd58a8632ddeb4d5c60d7c43)

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/879c13bfd1df34f61092abfcdcaae931ad30b650..861c712028d4b8b5cd58a8632ddeb4d5c60d7c43)

**Comment by facebook-github-bot-6** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/861c712028d4b8b5cd58a8632ddeb4d5c60d7c43..f329fcc379b07f10e270559b9079669e77f5f9c3)

File: Libraries/Animated/src/AnimatedImplementation.js (hunk 6 above)
Comment (kmagiera) [inline Libraries/Animated/src/AnimatedImplementation.js]: We should perhaps add to the documentation that creating event in render method is not the most effective way

**Comment by kmagiera** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-0** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3759844) to review internal test results.

**Comment by facebook-github-bot-7** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by brentvatne** [comment]
@bestander - any more info for why this import failed? :) :bow:

**Comment by bestander** [comment]
A jest test fails

```
RelayPersistedQueryChecker discovered that the persisted queries need to be
rebuilt for the following reason:

    `ScrollView` has no propType for native prop `RCTScrollView.accessibilityLabel` of native type `String`
If you haven't changed this prop yourself, this usually means that your versions of the native code and JavaScript code are out of sync. Updating both should make this error go away.

You can rebuild the persisted queries by running the following commands:
```

**Comment by janicduplessis** [comment]
@bestander Could you have someone look into it, looks pretty related to FB infra.

**Comment by janicduplessis** [comment]
It does look unrelated, unless adding Animated.ScrollView caused this for some reason so maybe just retrying could also work.

**Comment by facebook-github-bot-8** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/f329fcc379b07f10e270559b9079669e77f5f9c3..b4fde7cebebc4577e3eca61e6b81c2ab424c077a) - [changes since last import](https://github.com/facebook/react-native/pull/9253/files/f329fcc379b07f10e270559b9079669e77f5f9c3..b4fde7cebebc4577e3eca61e6b81c2ab424c077a)

**Comment by janicduplessis** [comment]
ping @bestander :)

**Comment by bestander** [comment]
I'll be back at a proper computer next week, vacation

**Comment by janicduplessis** [comment]
@bestander 👍 Thanks! Enjoy your vacations :)

**Comment by bestander** [comment]
Asked @kassens to help, he has some context on the failing test

**Comment by foghina** [comment]
Ping @bestander @kassens - do you guys need any help?

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

Messaging @kassens.

**Comment by bestander** [comment]
I am working on it actually

**Comment by bestander** [comment]
One test was failing

**Comment by facebook-github-bot-9** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/b4fde7cebebc4577e3eca61e6b81c2ab424c077a..69cff194103ca94ebe9b7d3fda669ce32939fbc7) - [changes since last import](https://github.com/facebook/react-native/pull/9253/files/f329fcc379b07f10e270559b9079669e77f5f9c3..69cff194103ca94ebe9b7d3fda669ce32939fbc7)

**Comment by janicduplessis** [comment]
Rebased to fix conflict, @bestander let me know if there is anything I can do to help.

**Comment by bestander** [comment]
Thanks, @janicduplessis, I posted this in the group https://www.facebook.com/groups/reactnativeoss/permalink/1588313934798568/?comment_id=1588884458074849&comment_tracking=%7B%22tn%22%3A%22R%22%7D

TL;DR: there is some weird test infra issue that blocks the merge.
It is not related to the code, I just don't want to disable tests and would rather fix it.
Planning to poke the people with more context and land it next week

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/4efd22846ecdb5cd451468d79ce0e8224a689cbf..e6668e43fd909a9e170df7d868c54ee6f21ed0a5) - [changes since last import](https://github.com/facebook/react-native/pull/9253/files/f329fcc379b07f10e270559b9079669e77f5f9c3..e6668e43fd909a9e170df7d868c54ee6f21ed0a5)

**Comment by facebook-github-bot-4** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9253/files/e6668e43fd909a9e170df7d868c54ee6f21ed0a5..4efd22846ecdb5cd451468d79ce0e8224a689cbf) - [changes since last import](https://github.com/facebook/react-native/pull/9253/files/f329fcc379b07f10e270559b9079669e77f5f9c3..4efd22846ecdb5cd451468d79ce0e8224a689cbf)

**Comment by foghina** [comment]
@facebook-github-bot shipit

**Comment by foghina** [comment]
Just importing it again, see if I can fix the test somehow.

**Comment by facebook-github-bot-4** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3759844) to review internal test results.

**Comment by facebook-github-bot-9** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by nihgwu** [comment]
@janicduplessis how can I use this in a `ListView`, I've just tried but failed with warning __Invalid prop 'onScroll' of type 'object' supplied to 'ListView', expected 'function'__

**Comment by janicduplessis** [comment]
To use it with ListView you'll need to create an AnimatedListView using `Animated.createAnimatedComponent`.

**Comment by nihgwu** [comment]
I did that but still failed, have you tried it?

???? iPhone

? 2016?9?19??01:25?Janic Duplessis <notifications@github.com<mailto:notifications@github.com>> ???


To use it with ListView you'll need to create an AnimatedListView using Animated.createAnimatedComponent.

-

**Comment by janicduplessis** [comment]
Ok, I think I know why. Does it work if you add
``` js 
  getScrollableNode: function() {
    if (this._scrollComponent && this._scrollComponent.getScrollableNode) {
      return this._scrollComponent.getScrollableNode();
    }
  },
```
in ListView.js

**Comment by nihgwu** [comment]
@janicduplessis It's definitely awesome, working so great, say goodbye to bad scrolling effects, thank you again, wish this can be merged soon

**Comment by foghina** [comment]
BOOM.

**Comment by nihgwu** [comment]
@janicduplessis what about to add support for `ListView`

**Comment by janicduplessis** [comment]
I'll make another PR for the ListView fix.
